plugins/out_clp_s3/flush/flush.go

```diff
@@ -21,111 +22,226 @@ import (
-			Timestamp:  ffi.EpochTimeMs(timestamp.UnixMilli()),
+		tag, err = NewTag(
+			tagKey,
+			ctx.Config.TimeZone,
+			size,
+			ctx.Config.UseDiskBuffer,
+			irBuf,
+			zstdBuf,
+		)
+
+		if err != nil {
+			return output.FLB_RETRY, fmt.Errorf("error creating tag: %w", err)
 		}
-		logEvents = append(logEvents, event)
+		ctx.Tags[tagKey] = tag
 	}
 
-	var buf bytes.Buffer
-
-	zstdWriter, err := zstd.NewWriter(&buf)
+	err = tag.Writer.WriteIrZstd(logEvents)
 	if err != nil {
-		err = fmt.Errorf("error opening zstd writer: %w", err)
-		return output.FLB_RETRY, err
+		return output.FLB_ERROR, err
 	}
 
-	// IR buffer using bytes.Buffer internally, so it will dynamically grow if undersized. Using
-	// FourByteEncoding as default encoding.
-	irWriter, err := ir.NewWriterSize[ir.FourByteEncoding](length, ctx.Config.TimeZone)
+	readyToUpload, err := checkUploadCriteria(
+		tag,
+		ctx.Config.UseDiskBuffer,
+		ctx.Config.UploadSizeMb,
+	)
 	if err != nil {
-		err = fmt.Errorf("error opening IR writer: %w", err)
-		return output.FLB_RETRY, err
+		return output.FLB_ERROR, fmt.Errorf("error checking upload criteria: %w", err)
 	}
 
-	err = writeIr(irWriter, logEvents)
-	if err != nil {
-		err = fmt.Errorf("error while encoding IR: %w", err)
-		return output.FLB_ERROR, err
+	if readyToUpload {
+		err := FlushZstdToS3(tag, ctx)
+		if err != nil {
+			return output.FLB_ERROR, fmt.Errorf("error flushing Zstd buffer to s3: %w", err)
+		}
 	}
 
-	// Write zstd compressed IR to file.
-	_, err = irWriter.CloseTo(zstdWriter)
+	return output.FLB_OK, nil
+}
+
+// Creates a new tag containing a new [irzstd.IrZstdWriter].
+//
+// Parameters:
+//   - tagKey: Fluent Bit tag
+//   - timezone: Time zone of the log source
+//   - size: Byte length
+//   - useDiskBuffer: On/off for disk buffering
+//   - irBuffer: Buffer for IR
+//   - ZstdBuffer: Buffer for Zstd compressed IR
+//
+// Returns:
+//   - tag: Tag resources and metadata
+//   - err: Error creating new writer
+func NewTag(
+	tagKey string,
+	timezone string,
+	size int,
+	useDiskBuffer bool,
+	irBuffer io.ReadWriter,
+	zstdBuffer io.ReadWriter,
+) (*outctx.Tag, error) {
+	writer, err := irzstd.NewIrZstdWriter(timezone, size, useDiskBuffer, irBuffer, zstdBuffer)
 	if err != nil {
-		err = fmt.Errorf("error writting IR to buf: %w", err)
-		return output.FLB_RETRY, err
+		return nil, err
 	}
 
-	err = zstdWriter.Close()
+	tag := outctx.Tag{
+		Key:    tagKey,
+		Writer: writer,
+	}
+
+	return &tag, nil
+}
+
+// Sends Zstd buffer to s3 and reset writer and buffers for future uploads. Prior to upload,
+// IR buffer is flushed and IR/Zstd streams are terminated. The tags index is incremented on
+// successful upload.
+//
+// Parameters:
+//   - tag: Tag resources and metadata
+//   - ctx: Plugin context
+//
+// Returns:
+//   - err: Error creating closing writer, error with type assertion, error uploading to s3,
+//
+// error resetting writer
+func FlushZstdToS3(tag *outctx.Tag, ctx *outctx.S3Context) error {
```

Comment:
```suggestion
func flushZstdToS3(tag *outctx.Tag, ctx *outctx.S3Context) error {
```
This doesn't appear used outside `flush` can it be private?